pkg/cloudprovider/aws/controllers/notification/controller.go

```diff
@@ -0,0 +1,326 @@
+	"github.com/aws/karpenter/pkg/cloudprovider/aws/events"
+	"github.com/aws/karpenter/pkg/controllers/state"
+	"github.com/aws/karpenter/pkg/metrics"
+)
+
+type Action = string
+
+var Actions = struct {
+	CordonAndDrain,
+	Cordon,
+	NoAction Action
+}{
+	CordonAndDrain: "CordonAndDrain",
+	Cordon:         "Cordon",
+	NoAction:       "NoAction",
+}
+
+// Controller is the notification controller. It is not a standard controller-runtime controller in that it doesn't
+// have a reconcile method.
+type Controller struct {
+	kubeClient           client.Client
+	cluster              *state.Cluster
+	recorder             events.Recorder
+	clock                clock.Clock
+	provider             *aws.SQSProvider
+	instanceTypeProvider *aws.InstanceTypeProvider
+	parser               event.Parser
+
+	infraController *infrastructure.Controller
+	backoff         *backoff.ExponentialBackOff
+}
+
+// pollingPeriod that we go to the SQS queue to check if there are any new events
+const pollingPeriod = 2 * time.Second
+
+func NewController(ctx context.Context, kubeClient client.Client, clk clock.Clock,
+	recorder events.Recorder, cluster *state.Cluster, sqsProvider *aws.SQSProvider,
+	instanceTypeProvider *aws.InstanceTypeProvider, infraController *infrastructure.Controller,
+	startAsync <-chan struct{}) *Controller {
+
+	c := &Controller{
+		kubeClient:           kubeClient,
+		cluster:              cluster,
+		recorder:             recorder,
+		clock:                clk,
+		provider:             sqsProvider,
+		instanceTypeProvider: instanceTypeProvider,
+		parser:               aggregatedparser.NewAggregatedParser(aggregatedparser.DefaultParsers...),
+		infraController:      infraController,
+		backoff:              newBackoff(clk),
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("notification"))
+	logging.FromContext(ctx).Infof("Starting controller")
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			return
+		case <-startAsync:
+			c.run(ctx)
+		}
+	}()
+
+	return c
+}
+
+func newBackoff(clk clock.Clock) *backoff.ExponentialBackOff {
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = time.Second * 2
+	b.MaxElapsedTime = time.Minute * 30
+	b.Clock = clk
+	return b
+}
+
+func (c *Controller) run(ctx context.Context) {
+	defer func() {
+		logging.FromContext(ctx).Infof("Shutting down")
+	}()
+	for {
+		<-c.infraController.Ready() // block until the infrastructure is up and ready
+		err := c.PollSQS(ctx)
+		if err != nil {
+			logging.FromContext(ctx).Errorf("Handling notification messages from SQS queue, %v", err)
+			select {
+			case <-ctx.Done():
+				return
+			case <-c.clock.After(c.backoff.NextBackOff()):
+				continue
+			}
+		}
+		c.backoff.Reset() // We succeeded so reset the backoff period
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.clock.After(pollingPeriod):
+		}
+	}
+}
+
+func (c *Controller) PollSQS(ctx context.Context) error {
```

Comment:
I disagree with this pretty strongly, primarily because `controller-runtime` reconcilers are meant to respond to Kubernetes object events to trigger reconciliation. 

This is evident from the fact that the `Reconcile` method takes in a `reconcile.Request` which is a wrapper around `types.NamespacedName`. It isn't well-suited for periodic polling and feels unnatural to me.

The way I would bucketize it is:

1. If you need to reconcile based on some Kubernetes object or some object event, use `controller-runtime`
2. Otherwise, write a periodic poller using a controller written in the style of `pkg/consolidation/controller`

I think this is a reasonable pattern to use with the project going forward and I, personally, don't find it any less readable. In fact, I would argue that it's more readable based on the context.

Reply:
Had a separate discussion today offline, decision was made to migrate to `controller-runtime`-based controllers